### PR TITLE
[RFR] Use a new algorithm to improve the solver

### PR DESCRIPTION
**Pull request by Luwangel**
Use a breadth first algorithm

### CODE REVIEW

File: src/events/game_listener.go [hunk 1]
```diff
@@ -8,6 +8,12 @@ import (
 	"time"
 )
 
+const TIME_SLEEP time.Duration = 800
+
+func wait() {
```
Comment (jdemangeon) [inline src/events/game_listener.go]: This function is useless IMO

File: src/game/shuffler.go [hunk 2]
```diff
@@ -6,27 +6,41 @@ import (
...
+				}
 
-			tileToMove = ChoiceCoordsNoSeed(movableTiles)
-			gridShuffled, _ = Move(gridShuffled, tileToMove)
-			movableTiles, err = ListMovableTilesWithoutGoingBack(gridShuffled, gridShuffled[tileToMove.Y][tileToMove.X])
-			count++
+				tileToMove = ChoiceCoordsNoSeed(movableTiles)
+				gridShuffled, _ = Move(gridShuffled, tileToMove)
+				movableTiles, err = ListMovableTilesWithoutGoingBack(gridShuffled, gridShuffled[tileToMove.Y][tileToMove.X])
+				count++
```
Comment (jdemangeon) [inline src/game/shuffler.go]: What's count ? Can you give it a more relevant name?

File: src/game/suggest.go [hunk 3]
```diff
@@ -138,3 +155,104 @@ func SolvePuzzle(shuffledGrid Grid, solvedGrid Grid, timeout chan bool) ([]Coord
...
+			previousValue := node.Grid[previousMoveCoords.Y][previousMoveCoords.X]
+			possibleMoves, _ := ListMovableTilesWithoutGoingBack(node.Grid, previousValue)
+			for _, neighbor := range BuildNeighborList(node, solvedGrid, possibleMoves) {
+				sublist = append(sublist, neighbor)
+			}
+		}
+	}
+	listChan <- sublist
+}
+
```
Comment (jdemangeon) [inline src/game/suggest.go:189]: s/findGridResolvedInList/findResolvedGridInList

File: src/game/suggest.go [hunk 4]
```diff
@@ -138,3 +155,104 @@ func SolvePuzzle(shuffledGrid Grid, solvedGrid Grid, timeout chan bool) ([]Coord
...
+func SolvePuzzleDWithPreviousMove(shuffledGrid Grid, solvedGrid Grid, lastMove Coords) ([]Coords, error) {
+	previousValue := shuffledGrid[lastMove.Y][lastMove.X]
+	possibleMoves, _ := ListMovableTilesWithoutGoingBack(shuffledGrid, previousValue)
+
+	return solvePuzzleD(shuffledGrid, solvedGrid, possibleMoves)
+}
+
+func solvePuzzleD(shuffledGrid Grid, solvedGrid Grid, possibleMoves []Coords) ([]Coords, error) {
```
Comment (jdemangeon) [inline src/game/suggest.go:220]: Can you try to split this function to make it clearer?

Comment (jdemangeon) [reply src/game/suggest.go:220]: Moreover it'll help you to test because this function is complex (and important)

Comment (Luwangel) [reply src/game/suggest.go:220]: I've split a little bit this function. I think the whole file needs to be split in a specific submodule to distinct the heuristic, the priority list etc.

**Review by jdemangeon — APPROVED** [review]

**Pull request merged**